### PR TITLE
Unit test and fixed bug

### PR DESCRIPTION
**Pull request by kketernality**
This pull request contains the following updates:

### Unit Test

1. Implemented unit test using Catch2 [[1]](https://github.com/catchorg/Catch2) and ported suitable tests from the original Python repository [[2]](https://github.com/seatgeek/fuzzywuzzy/blob/master/test_fuzzywuzzy.py). The unit test files locate in the `test` folder as before.
2. Updated `CMakeLists.txt` and `README.md` to provide instruction of building and running unit test.

### Fixed Bugs

- [v] The example from Readme gave the wrong result. The cause is that the `full_process` is default on and will strip the tailing exclamation mark. It was fixed by altering the default value of `full_process` to `false`.
```
ReadmeExample
FAILED:
      REQUIRE( 97 == fuzz::ratio("this is a test", "this is a test!") );
EXPANSION:
      97 == 100 
```
- [v] Comparison of two empty strings should give 100% similarity as in the original Python implementation. Fixed by adding additional logic in `ratio()` and `partial_ratio()`.
```
RatioTest
      testEmptyStringsScore100
FAILED:
      REQUIRE( 100 == fuzz::ratio("", "") )
EXPANSION:
      100 == 0
```
- [v] The `partial_ratio()` of two identical string is 0% similarity which should be 100% undoubtedly. The detailed description is in the `fuzzywuzzy.cpp`. This was fixed by a temporary solution to compare if two strings is identical or not. 
```
RatioTest
      testPartialRatio
FAILED:
      REQUIRE( 100 == fuzz::partial_ratio("new york mets", "new york mets") )
EXPANSION:
      100 == 0
```

### CODE REVIEW

**Comment by tmplt** [comment]
Thanks for taking the time to port unit tests over! I'll see about going over the commits the coming week.

**Comment by kketernality** [comment]
Just let me know if there is anything goes wrong or needs to be revisited ^_^

File: test/main.cpp [hunk 1]
```diff
@@ -1,13 +1,15 @@
...
-}
+//int main()
+//{
+//    const string a = "I'm in your mind", b = "I'm in your mind fuzz";
+//    const string c = "fuzzy wuzzy was a bear", d = "wuzzy fuzzy was a bear";
+//
+//    std::cout << fuzz::ratio(a, b) << '\n';
+//    std::cout << fuzz::partial_ratio(a, b) << '\n';
+//    std::cout << fuzz::token_sort_ratio(c, d) << '\n';
+//}
```
Comment (tmplt) [inline test/main.cpp]: This commented code can be removed.

File: test/test_fuzzywuzzy.cpp [hunk 2]
```diff
@@ -0,0 +1,430 @@
...
+        REQUIRE( 100 == fuzz::partial_ratio(s1, s3) );
+    }
+
+    SECTION("testTokenSortRatio")
+    {
+        REQUIRE( 100 == fuzz::token_sort_ratio(s1, s1a) );
+    }
+
+    /*
+    SECTION("testPartialTokenSortRatio")
```
Comment (tmplt) [inline test/test_fuzzywuzzy.cpp]: Any reason this is commented out?

File: test/test_fuzzywuzzy.cpp [hunk 3]
```diff
@@ -0,0 +1,430 @@
...
+
+    SECTION("testGetBestChoice4")
+    {
+        const std::string query = "chicago cubs vs new york mets";
+        auto best = fuzz::extractOne(query, baseball_strings);
+        REQUIRE( best[0].first == baseball_strings[0] );
+    }
+
+    /*
+    SECTION("testWithProcessor")
```
Comment (tmplt) [inline test/test_fuzzywuzzy.cpp]: ditto.

File: test/test_fuzzywuzzy.cpp [hunk 4]
```diff
@@ -0,0 +1,430 @@
...
+    {
+        const std::vector<std::string> choices = {
+            "new york mets vs chicago cubs",
+            "chicago cubs at new york mets",
+            "atlanta braves vs pittsbugh pirates",
+            "new york yankees vs boston red sox"
+        };
+
+        /*
+        const std::unordered_map<int, std::string> choices_dict = {
```
Comment (tmplt) [inline test/test_fuzzywuzzy.cpp]: ditto.

File: README.md [hunk 5]
```diff
@@ -11,6 +11,25 @@ and been wrapped around some C++ code.
...
+cd build
+cmake ..
+make -j$(nproc) # Multi-threaded compilation
+```
+
+Run the unit test of this library
+
+```bash
+cd build
+make test
```
Comment (tmplt) [inline README.md]:
```
$ make test
Running tests...
Test project /home/tmplt/code/fuzzywuzzy/build
    Start 1: main_test
1/1 Test #1: main_test ........................   Passed    0.01 sec

100% tests passed, 0 tests failed out of 1
```
this PR contains a lot more than a single test, so I'm guessing that the empty binary produced by `test/main.cpp` is executed here instead of the wanted `test/test_*.cpp`?

Comment (kketernality) [reply README.md:30]: The test is linked and executed together. The Catch2 framework generates the main and executes all the ‘TEST_CASE’. 
The ‘make test’ command simply executes the unit test executable.

File: test/test_fuzzywuzzy.cpp (hunk 2 above)
Comment (kketernality) [reply test/test_fuzzywuzzy.cpp:139]: The function ‘void partial_token_sort_ratio()’ is not implemented yet.

File: test/test_fuzzywuzzy.cpp (hunk 3 above)
Comment (kketernality) [reply test/test_fuzzywuzzy.cpp:286]: The C++ implementation does not accept this argument yet. Python can do it because it doesn’t care the type.

File: test/test_fuzzywuzzy.cpp (hunk 4 above)
Comment (kketernality) [reply test/test_fuzzywuzzy.cpp:314]: Same as the above one.

**Comment by kketernality** [comment]
> I'm not partial to adding a ~17kloc library header. Could you source `catch.hpp` from a submodule instead?

Okay, I think make it as an installed package or something likewise is more proper. Will be added in the next commit.

**Comment by kketernality** [comment]
@tmplt  

I had updated the required change. Please review the code changes in this new commit.
